src/sisl/utils/misc.py

```diff
@@ -220,6 +221,71 @@ def str_spec(name):
     return "{".join(lname[:-1]), lname[-1]
 
 
+IterableAny = Iterator[Any]
+IterableInstantiation = Callable[..., IterableAny]
+
+
+class Listify:
+    """Convert arguments to an iterable-like (any iterable, default to `list`)
+
+    It provides also an easy mechanism for "piping" content to
+    a function call (for better readability).
+
+    Parameters
+    ----------
+    cls:
+        the default list-like object to cast it to
+
+    Examples
+    --------
+    >>> listify = Listify()
+    >>> 1 | listify
+    [1]
+    >>> Listify(tuple)(1)
+    (1,)
+
+    It can greatly improve readability with `map` constructs:
+    >>> map(lambda x, [1]) | listify
+    [1]
+
+    Notes
+    -----
+    If using this to convert to `tuple` instances ``Iterfy(tuple)``,
+    please do note the problems of using a tuple as indices for
+    `numpy.ndarray` objects.
+    """
+
+    __slots__ = ("_cls",)
+
+    # Ensures that numpy function calls won't happen!
+    # Just a higher priority than *any* numpy arrays and subclasses.
+    __array_priority__ = 1000000
+
+    def __init__(self, cls: IterableInstantiation = list):
+        self._cls = cls
+
+    def __call__(
+        self, arg: Any, cls: Optional[IterableInstantiation] = None
+    ) -> IterableAny:
+        if cls is None:
+            cls = self._cls
+        if isinstance(arg, Iterable):
+            if isinstance(arg, cls):
+                return arg
+            return cls(arg)
+        return cls([arg])
+
+    def __ror__(self, arg: Any) -> IterableAny:
+        """Allow piping of function calls"""
+        return self(arg)
```

Comment:
Is this a good idea? 

Is `|` used for piping elsewhere with python? 

In the description you say:

> It provides also an easy mechanism for "piping" content to
>     a function call (for better readability).

It is better readability if `|` would be used for piping in general in python, but since it is used for `or` operations I think it can be more confusing than readable :sweat_smile:

Reply:
Hmm, I think here it is quite clear, have you seen this package https://pypi.org/project/pipe/
I think the expressiveness without parenthesis makes it more clear, although I can see it a bit confusing if you are not familiar with pipes on Unix. In any case it is more used internally here.

Reply:
I searched for pipes in python and I found that package. But calling a function with `|` is not something that is considered in the specification of the syntax.

Even if you are used to `|` meaning "pipe" in bash, it's still confusing to find it in python. I wouldn't think about it if I see it in python code. I would think that listify is some kind of fallback if the expression on the left is false 😅

Reply:
ha, I see. I think I'll try and keep it, the amount of parenthesis gets heavily reduced with this ;)